lib/stm32/l4/rcc.c

```diff
@@ -341,6 +380,83 @@ uint32_t rcc_system_clock_source(void)
 	return (RCC_CFGR >> RCC_CFGR_SWS_SHIFT) & RCC_CFGR_SWS_MASK;
 }
 
+/**
+ * Setup clocks to run from PLL.
+ *
+ * The arguments provide the pll source, multipliers, dividers, all that's
+ * needed to establish a system clock.
+ *
+ * @param clock clock information structure.
+ */
+void rcc_clock_setup_pll(const struct rcc_clock_scale *clock)
+{
+  /* Enable internal high-speed oscillator (HSI16). */
+  rcc_osc_on(RCC_HSI16);
+  rcc_wait_for_osc_ready(RCC_HSI16);
+  
+  /* Select HSI16 as SYSCLK source. */
+  rcc_set_sysclk_source(RCC_PLLCFGR_PLLSRC_HSI16);
+  
+  /* Enable external high-speed oscillator (HSE). */
+  if (clock->pll_source == RCC_PLLCFGR_PLLSRC_HSE) {
+    rcc_osc_on(RCC_HSE);
+    rcc_wait_for_osc_ready(RCC_HSE);
+  }
+  
+  /* Set the VOS scale mode */
+  rcc_periph_clock_enable(RCC_PWR);
+  pwr_set_vos_scale(clock->voltage_scale);
+  
+  /*
+   * Set prescalers for AHB, ADC, APB1, APB2.
+   * Do this before touching the PLL (TODO: why?).
+   */
+  rcc_set_hpre(clock->hpre);
+  rcc_set_ppre1(clock->ppre1);
+  rcc_set_ppre2(clock->ppre2);
+  
+  /* Disable PLL oscillator before changing its configuration. */
+  rcc_osc_off(RCC_PLL);
+  
+  /* Configure the PLL oscillator. */
+  rcc_set_main_pll(clock->pll_source, clock->pllm, clock->plln,
+                         clock->pllp, clock->pllq, clock->pllr);
+  
+  /* Enable PLL oscillator and wait for it to stabilize. */
+  rcc_osc_on(RCC_PLL);
+  rcc_wait_for_osc_ready(RCC_PLL);
+  
+  /* Configure flash settings. */
+  if (clock->flash_config & FLASH_ACR_DCEN) {
+    flash_dcache_enable();
+  } else {
+    flash_dcache_disable();
+  }
+  if (clock->flash_config & FLASH_ACR_ICEN) {
+    flash_icache_enable();
+  } else {
+    flash_icache_disable();
+  }
+  flash_set_ws(clock->flash_config);
+  
+  /* Select PLL as SYSCLK source. */
+  rcc_set_sysclk_source(RCC_CFGR_SW_PLL);
+  
+  /* Wait for PLL clock to be selected. */
+  rcc_wait_for_sysclk_status(RCC_PLL);
+  
+  return;
```

Comment:
wat?

Reply:
Thank you for taking your time with this. I'm a real newbie when it comes to VCS and contributing to others' code. This function I got line for line from lib/stm32/f4/rcc.c:789. I definitely forgot to make my formatting consistent. Should I fix the formatting and add another commit or should I alter the history? (again, I'm not perfectly familiar with the etiquette)

Reply:
more commits is fine if the end result is just one feature.  I can just squash it when I merge it.    Where I really ask for separate commits is where you have multiple separate functional changes.